src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+                        {
+                            _state++;
+                            _current = _enumerator.Current;
+                            return true;
+                        }
+
+                        break;
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public IPartition<TSource> Skip(int count)
+            {
+                int minIndex = _minIndexInclusive + count;
+                return (uint)minIndex > (uint)_maxIndexInclusive ? EmptyPartition<TSource>.Instance : new EnumerablePartition<TSource>(_source, minIndex, _maxIndexInclusive);
+            }
+
+            public IPartition<TSource> Take(int count)
+            {
+                int maxIndex = _minIndexInclusive + count - 1;
+                return (uint)maxIndex >= (uint)_maxIndexInclusive ? this : new EnumerablePartition<TSource>(_source, _minIndexInclusive, maxIndex);
+            }
+
+            public TSource TryGetElementAt(int index, out bool found)
+            {
+                if ((uint)index <= (uint)(_maxIndexInclusive - _minIndexInclusive)) // Fail fast if possible.
+                {
+                    using (IEnumerator<TSource> en = _source.GetEnumerator())
+                    {
+                        if (SkipBeforeFirst(en) && TakeBefore(index, en) && en.MoveNext())
+                        {
+                            found = true;
+                            return en.Current;
+                        }
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource TryGetFirst(out bool found)
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        found = true;
+                        return en.Current;
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource TryGetLast(out bool found)
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en))
+                    {
+                        return TryGetLastTaken(en, out found);
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource[] ToArray()
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        var builder = new LargeArrayBuilder<TSource>(initialize: true);
+                        int toTake = _maxIndexInclusive - _minIndexInclusive + 1;
```

Comment:
Note: This can overflow if `_maxIndexInclusive` is int.MaxValue, and `_minIndexInclusive` is 0 (which happens for `e.Skip(0)`). It doesn't matter though, since we decrement the value right away.